File: Core Data/WWDC 2015/Whats_new_in_core_data/iPhoneCoreDataRecipes/Classes/RecipeTableViewCell.m
/*
 Copyright (C) 2017 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 A table view cell that displays information about a Recipe.  It uses individual subviews of its content view to show the name, picture, description, and preparation time for each recipe.  If the table view switches to editing mode, the cell reformats itself to move the preparation time off-screen, and resizes the name and description fields accordingly.
 */

#import "RecipeTableViewCell.h"

@interface RecipeTableViewCell ()

@property (nonatomic, strong) UIImageView *recipeImageView;
@property (nonatomic, strong) UILabel *nameLabel;
@property (nonatomic, strong) UILabel *overviewLabel;
@property (nonatomic, strong) UILabel *prepTimeLabel;

@property (NS_NONATOMIC_IOSONLY, readonly) CGRect _imageViewFrame;
@property (NS_NONATOMIC_IOSONLY, readonly) CGRect _nameLabelFrame;
@property (NS_NONATOMIC_IOSONLY, readonly) CGRect _descriptionLabelFrame;
@property (NS_NONATOMIC_IOSONLY, readonly) CGRect _prepTimeLabelFrame;

@end


#pragma mark -

@implementation RecipeTableViewCell

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    
	if (self = [super initWithCoder:aDecoder]) {
        _recipeImageView = [[UIImageView alloc] initWithFrame:CGRectZero];
		self.recipeImageView.contentMode = UIViewContentModeScaleAspectFit;
        [self.contentView addSubview:self.recipeImageView];
        
        _overviewLabel = [[UILabel alloc] initWithFrame:CGRectZero];
        self.overviewLabel.font = [UIFont systemFontOfSize:12.0];
        self.overviewLabel.textColor = [UIColor darkGrayColor];
        self.overviewLabel.highlightedTextColor = [UIColor whiteColor];
        [self.contentView addSubview:self.overviewLabel];
        
        _prepTimeLabel = [[UILabel alloc] initWithFrame:CGRectZero];
        self.prepTimeLabel.textAlignment = NSTextAlignmentRight;
        self.prepTimeLabel.font = [UIFont systemFontOfSize:12.0];
        self.prepTimeLabel.textColor = [UIColor blackColor];
        self.prepTimeLabel.highlightedTextColor = [UIColor whiteColor];
		self.prepTimeLabel.minimumScaleFactor = 7.0;
		self.prepTimeLabel.lineBreakMode = NSLineBreakByTruncatingTail;
        [self.contentView addSubview:self.prepTimeLabel];
        
        _nameLabel = [[UILabel alloc] initWithFrame:CGRectZero];
        self.nameLabel.font = [UIFont boldSystemFontOfSize:14.0];
        self.nameLabel.textColor = [UIColor blackColor];
        self.nameLabel.highlightedTextColor = [UIColor whiteColor];
        [self.contentView addSubview:self.nameLabel];
    }
    
    return self;
}

// To save space, the prep time label disappears during editing.
- (void)layoutSubviews {
    
    [super layoutSubviews];
	
    self.recipeImageView.frame = [self _imageViewFrame];
    self.nameLabel.frame = [self _nameLabelFrame];
    self.overviewLabel.frame = [self _descriptionLabelFrame];
    self.prepTimeLabel.frame = [self _prepTimeLabelFrame];
    if (self.editing) {
        self.prepTimeLabel.alpha = 0.0;
    } else {
        self.prepTimeLabel.alpha = 1.0;
    }
}


#define IMAGE_SIZE          42.0
#define EDITING_INSET       10.0
#define TEXT_LEFT_MARGIN    8.0
#define TEXT_RIGHT_MARGIN   5.0
#define PREP_TIME_WIDTH     80.0

// Returns the frame of the various subviews -- these are dependent on the editing state of the cell.
- (CGRect)_imageViewFrame {
    
    if (self.editing) {
        return CGRectMake(EDITING_INSET, 0.0, IMAGE_SIZE, IMAGE_SIZE);
    }
	else {
        return CGRectMake(0.0, 0.0, IMAGE_SIZE, IMAGE_SIZE);
    }
}

- (CGRect)_nameLabelFrame {
    
    if (self.editing) {
        return CGRectMake(IMAGE_SIZE + EDITING_INSET + TEXT_LEFT_MARGIN, 4.0, self.contentView.bounds.size.width - IMAGE_SIZE - EDITING_INSET - TEXT_LEFT_MARGIN, 16.0);
    }
	else {
        return CGRectMake(IMAGE_SIZE + TEXT_LEFT_MARGIN, 4.0, self.contentView.bounds.size.width - IMAGE_SIZE - TEXT_RIGHT_MARGIN * 2 - PREP_TIME_WIDTH, 16.0);
    }
}

- (CGRect)_descriptionLabelFrame {
    
    if (self.editing) {
        return CGRectMake(IMAGE_SIZE + EDITING_INSET + TEXT_LEFT_MARGIN, 22.0, self.contentView.bounds.size.width - IMAGE_SIZE - EDITING_INSET - TEXT_LEFT_MARGIN, 16.0);
    }
	else {
        return CGRectMake(IMAGE_SIZE + TEXT_LEFT_MARGIN, 22.0, self.contentView.bounds.size.width - IMAGE_SIZE - TEXT_LEFT_MARGIN, 16.0);
    }
}

- (CGRect)_prepTimeLabelFrame {
    
    CGRect contentViewBounds = self.contentView.bounds;
    return CGRectMake(contentViewBounds.size.width - PREP_TIME_WIDTH - TEXT_RIGHT_MARGIN, 4.0, PREP_TIME_WIDTH, 16.0);
}


#pragma mark - Recipe set accessor

- (void)setRecipe:(Recipe *)newRecipe {
    
    if (newRecipe != _recipe) {
        _recipe = newRecipe;
	}
	self.recipeImageView.image = _recipe.thumbnailImage;
	self.nameLabel.text = (_recipe.name.length > 0) ? _recipe.name : @"-";
	self.overviewLabel.text = (_recipe.overview != nil) ? _recipe.overview : @"-";
	self.prepTimeLabel.text = (_recipe.prepTime != nil) ? _recipe.prepTime : @"-";
}

@end

File: Core Data/Core Data Sample Codes/iPhoneCoreDataRecipes/Recipes_Swift_3_Updated/Recipes_Swift/Table Cells/RecipeTableViewCell.swift
/*
  RecipeTableViewCell.swift
  Recipes_Swift

  Created by Seyed Samad Gholamzadeh on 1/26/18.
  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.

Abstract:
A table view cell that displays information about a Recipe.  It uses individual subviews of its content view to show the name, picture, description, and preparation time for each recipe.  If the table view switches to editing mode, the cell reformats itself to move the preparation time off-screen, and resizes the name and description fields accordingly.
*/


import UIKit

private let imageSize: CGFloat = 42.0
private let editingInset: CGFloat = 10.0
private let textLeftMargin: CGFloat = 8.0
private let textRightMargin: CGFloat = 5.0
private let prepTimeWidth: CGFloat = 80.0

class RecipeTableViewCell: UITableViewCell {

	var recipeImageView: UIImageView!
	var nameLabel: UILabel!
	var overviewLabel: UILabel!
	var prepTimeLabel: UILabel!
	
	//MARK: - Recipe set accessor
	var recipe: Recipe! {
		didSet {
			if recipe != oldValue {
				self.recipeImageView.image = recipe.thumbnailImage
				self.nameLabel.text = recipe.name != nil && !recipe.name!.isEmpty ? recipe.name! : "-"
				self.overviewLabel.text = recipe.overview != nil ? recipe.overview : "-"
				self.prepTimeLabel.text = recipe.prepTime != nil ? recipe.prepTime : "-"
			}
		}
	}
	

	var imageViewFrame: CGRect {
		if self.isEditing {
			return CGRect(x: editingInset, y: 0, width: imageSize, height: imageSize)
		}
		else {
			return CGRect(x: 0, y: 0, width: imageSize, height: imageSize)
		}
	}
	
	var nameLabelFrame: CGRect {
		if self.isEditing {
			return CGRect(x: imageSize + editingInset + textLeftMargin, y: 4.0, width: self.contentView.bounds.size.width - imageSize - editingInset - textLeftMargin, height: 16.0)
		}
		else {
			return CGRect(x: imageSize + textLeftMargin, y: 4.0, width: self.contentView.bounds.size.width - imageSize - textRightMargin*2 - prepTimeWidth, height: 16.0)
		}
	}
	
	var descriptionLabelFrame: CGRect {
		if self.isEditing {
			return CGRect(x: imageSize + editingInset + textLeftMargin, y: 22.0, width: self.contentView.bounds.size.width - imageSize - editingInset - textLeftMargin, height: 16.0)
		}
		else {
			return CGRect(x: imageSize + textLeftMargin, y: 22.0, width: self.contentView.bounds.size.width - imageSize - textLeftMargin, height: 16.0)
		}
	}
	
	var prepTimeLabelFrame: CGRect {
		let contentViewBounds = self.contentView.bounds
		return CGRect(x: contentViewBounds.size.width - prepTimeWidth - textRightMargin, y: 4.0, width: prepTimeWidth, height: 16.0)
	}

	
	required init?(coder aDecoder: NSCoder) {
		super.init(coder: aDecoder)
		
		recipeImageView = UIImageView(frame: .zero)
		self.recipeImageView.contentMode = .scaleAspectFit
		self.contentView.addSubview(self.recipeImageView)
		
		self.overviewLabel = UILabel(frame: .zero)
		self.overviewLabel.font = UIFont.systemFont(ofSize: 12.0)
		self.overviewLabel.textColor = UIColor.darkGray
		self.overviewLabel.highlightedTextColor = .white
		self.contentView.addSubview(self.overviewLabel)
		
		self.prepTimeLabel = UILabel(frame: .zero)
		self.prepTimeLabel.textAlignment = .right
		self.prepTimeLabel.font = UIFont.systemFont(ofSize: 12.0)
		self.prepTimeLabel.textColor = .white
		self.prepTimeLabel.minimumScaleFactor = 7.0
		self.prepTimeLabel.lineBreakMode = .byTruncatingTail
		self.contentView.addSubview(self.prepTimeLabel)
		
		self.nameLabel = UILabel(frame: .zero)
		self.nameLabel.font = UIFont.boldSystemFont(ofSize: 14.0)
		self.nameLabel.textColor = .black
		self.nameLabel.highlightedTextColor = .white
		self.contentView.addSubview(self.nameLabel)
	}
	
	override func layoutSubviews() {
		super.layoutSubviews()
		
		self.recipeImageView.frame = self.imageViewFrame
		self.nameLabel.frame = self.nameLabelFrame
		self.overviewLabel.frame = self.descriptionLabelFrame
		self.prepTimeLabel.frame = self.prepTimeLabelFrame
		
		if self.isEditing {
			self.prepTimeLabel.alpha = 0.0
		}
		else {
			self.prepTimeLabel.alpha = 1.0
		}
	}
	
	
    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Configure the view for the selected state
    }

}
